api/resources_portal/views/grant_user.py

```diff
@@ -0,0 +1,73 @@
+from rest_framework import viewsets
+from rest_framework.exceptions import MethodNotAllowed
+from rest_framework.permissions import BasePermission, IsAuthenticated
+from rest_framework.response import Response
+
+from rest_framework_extensions.mixins import NestedViewSetMixin
+
+from resources_portal.models import Grant, GrantOrganizationAssociation, GrantUserAssociation, User
+from resources_portal.views.relation_serializers import UserRelationSerializer
+
+
+class OwnsGrant(BasePermission):
+    def has_permission(self, request, view):
+        grant = Grant.objects.get(pk=view.kwargs["parent_lookup_grants"])
+
+        return request.user in grant.users.all()
+
+
+class CanManageOrganizationUsers(BasePermission):
+    """Allows users to add grant-user relationships only if they have permission to manage
+    users on an organization associated with the grant and the user to add is in that organization."""
+
+    def has_permission(self, request, view):
+        grant = Grant.objects.get(pk=view.kwargs["parent_lookup_grants"])
+
+        for association in GrantOrganizationAssociation.objects.filter(grant=grant):
+            if (
+                request.user.has_perm(
+                    "add_members_and_manage_permissions", association.organization
```

Comment:
This permission is for users to add members to organizations and manage their permissions. This should not relate to grants at all.

I think grants is one area where we do want to force the PI to be involved. They are essentially only used for credit purposes, so generally most members of the organization won't need to worry about them, just the PI's who want to get credit for their grants sharing the resources they create.

So basically, only an owner of a grant should be allowed to modify it or add additional users to it.

Reply:
Gotcha. Currently, that is the case- all operations on this endpoint require the user to be an owner of the grant. This permission is to make sure that owners can only add users from organizations that are:
A. Associated with the grant
B. Organizations that the owner has some kind of permissions on

To remedy this issue, should I maybe create a new permission on organizations that's like "can_add_users_to_grants" or something like that?

Reply:
Alright, I think this has been addressed.